packages/components/src/style-provider/index.tsx

```diff
@@ -12,19 +11,27 @@ import * as uuid from 'uuid';
 import type { StyleProviderProps } from './types';
 
 const uuidCache = new Set();
+// Use a weak map so that when the container is detached it's automatically
+// dereferenced to avoid memory leak.
+const containerCacheMap = new WeakMap();
 
-const memoizedCreateCacheWithContainer = memoize(
-	( container: HTMLElement ) => {
-		// Emotion only accepts alphabetical and hyphenated keys so we just
-		// strip the numbers from the UUID. It _should_ be fine.
-		let key = uuid.v4().replace( /[0-9]/g, '' );
-		while ( uuidCache.has( key ) ) {
-			key = uuid.v4().replace( /[0-9]/g, '' );
-		}
-		uuidCache.add( key );
-		return createCache( { container, key } );
+const memoizedCreateCacheWithContainer = ( container: HTMLElement ) => {
+	if ( containerCacheMap.has( container ) ) {
+		return containerCacheMap.get( container );
 	}
-);
+
+	// Emotion only accepts alphabetical and hyphenated keys so we just
+	// strip the numbers from the UUID. It _should_ be fine.
+	let key = uuid.v4().replace( /[0-9]/g, '' );
+	while ( uuidCache.has( key ) ) {
+		key = uuid.v4().replace( /[0-9]/g, '' );
+	}
+	uuidCache.add( key );
+
+	const cache = createCache( { container, key } );
+	containerCacheMap.set( container, cache );
+	return cache;
+};
```

Comment:
Cc @sarayourfriend 

@dmsnell Weren't you talking about emotion going crazy? Does this fix it?

Reply:
@kevin940726 Is there any way we could tests this to prevent regressions? 🤔

Reply:
yeah I suppose this could be a source of the leak.

I never thought about `memize` being unbounded, though it looks like it does allow us to pass [a `maxSize` option as the second argument](https://github.com/aduth/memize/blob/master/index.js#L133) which will bound it such that it purges the least-recently-used entry when adding into the cache.

in other words, maybe it would be worth attempting this change first as this

```diff
 const memoizedCreatedCacheWithContainer = memoize(
 	( container: HTMLElement ) => { … },
+	{ maxSize: 1000 }
 }
```

Reply:
Well, WeakMap is a superior solution if it can be used. But yes, maybe we should always require a maxSize option for memoize https://github.com/WordPress/gutenberg/pull/53406#discussion_r1286726012

Reply:
are we defeating the purpose of the `WeakMap` by storing the container inside the cache?

`WeakMap` allows the release of the keys, but not the values. do we need to store a `WeakRef` to the container instead of the container itself?

does emotion allow this?

Reply:
I don't think `maxSize` prevents the memory leak. It simply adds an upper bound of what this cache can hold until it starts trimming the old ones. A `maxSize` of `1000` will still unnecessarily leak at most 1000 detached documents, which is still a lot. Making the value of `maxSize` smaller doesn't help either as it will just defeat the purpose of such cache.

AFAIK, `WeakMap` does allow values to be garbage collected. See this section of the [MDN doc](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap#why_weakmap):

> By contrast, in a `WeakMap`, a key object refers strongly to its contents as long as the key is not garbage collected, but weakly from then on. As such, a `WeakMap`:
>
> - does not prevent garbage collection, which eventually removes references to the key object
> - allows garbage collection of any values if their key objects are not referenced from somewhere other than a `WeakMap`

Reply:
> Is there any way we could tests this to prevent regressions? 

I believe we can test this some way or another in the performance test suite (c.c. @WunderBart). Though I would rather keep it in another PR since it's not trivial to do 😆.

Reply:
Excellent catch :+1: +100% to adding a rule to require max size on memoisation functions and/or to strongly suggest a fall back to a WeakMap. It is really a much better, harder to screw-up option for simple, single-argument cases like this.

Reply:
> WeakMap does allow values to be garbage collected

thanks for double-checking @kevin940726 - it was late when I was reading through the docs and I wasn't sure that holding the key as part of a value would remain weak.

so a fair amount of time has passed between when I wrote the previous sentence and now. I've done some testing in Safari and Chrome. testing was hard because at least in Chrome, the JavaScript engine seems to overlook small `WeakMap` instances, so I [took a suggestion](https://stackoverflow.com/a/72305723) and filled one up before manually triggering collection.

we have marginal insight here because the browser consoles show the contents of the `WeakMap`, which is something we can't do in code for obvious reasons. here's the summary of my testing:

 - yes they do let go of references to the objects if they are values of the `WeakMap`
 - but at least during some test runs it stubbornly held on to the references longer than I expected. this was worse when the values were `weakmap.set( o, { o, i } )` than when they were simpler as `weakmap.set( o, o )`
 - if the object is referenced in the value of some other key then it doesn't seem to let go. that is, `weakmap.set( b, [ a, b ] ); a = null;` leaves `a` allocated and it won't seem to free it until I also clear `b = null`
 - at times I would see this stubborn clinging lead to multiple copies of `c` in the `WeakMap` long after `a` and `b` were purged and long after each copy of `c` had been set as `c = null` and manual garbage collection was run. that is, the `WeakMap` can be leaky in this regard through multiple garbage collection cycles.

after a while of testing I found that things freed up more reliably when I tried. I'm guessing this could be related to the engine warming up.

in the process I found [`FinalizationRegistry()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry) which is really neat and I didn't know it existed. All of the debugging tools around garbage-collection though are intentionally blurry and non-deterministic. for a while I saw no finalization events, then they started appearing after several trial runs. browsers stress out to ensure we don't count on this behavior.

so take this as validation that we should be good here, but we also in general probably want to make sure we don't store the same object in two `WeakMap` values or they won't get freed until both `WeakMap` keys have been released.

Reply:
@dmsnell Thanks for the validation!

If you're using chrome devtools, clicking on the bin icon shown in my video will force garbage collection, which is consistent and instant in my testing. Another thing is that some browsers (chromium-based for instance) only collect garbage every 20 seconds even when you use some advanced memory API. So that might be the reason why you're getting the results late.

> but we also in general probably want to make sure we don't store the same object in two `WeakMap` values or they won't get freed until both `WeakMap` keys have been released.

Yep! This is working as expected because the value is still reachable by the other key. `WeakMap` is probably a rare advanced API that we don't normally use, but it fits perfectly in this case IMO.

Reply:
> clicking on the bin icon shown in my video will force garbage collection

right, but I saw plenty of cases where garbage collection didn't free the items in the `WeakMap`, which is something indicated in the docs too. they have their own heuristics for when to purge the entries and I was testing those out to see what it's like. so clicking on the button triggered garbage collection, but in many cases left the values in the `WeakMap` for some time - and for the most part, until the `WeakMap` was big enough that it was "heavy" and warranted purging.

maybe this appears in some of our measurements as retaining values too long, but those values aren't strong references so they could be purged at any time. the measurements might at times be misleading.